contracts/VolumeImbalanceRecorder.sol

```diff
@@ -0,0 +1,152 @@
+    tokenControlInfo[token] = TokenControlInfo(
+                                minimalRecordResolution,
+                                maxPerBlockImbalance,
+                                maxTotalImbalance);
+    SetTokenControlInfo( token,
+                         minimalRecordResolution,
+                         maxPerBlockImbalance,
+                         maxTotalImbalance );
+  }
+
+  function getTokenControlInfo( ERC20 token ) public view returns(uint,uint,uint) {
+    return (tokenControlInfo[token].minimalRecordResolution,
+            tokenControlInfo[token].maxPerBlockImbalance,
+            tokenControlInfo[token].maxTotalImbalance);
+  }
+
+  function getImbalanceInRange( ERC20 token, uint startBlock, uint endBlock ) public view returns(int buyImbalance ) {
+    // check the imbalance in the sliding window
+    require(startBlock <= endBlock);
+
+    buyImbalance = 0;
+
+    for(uint windowInd = 0 ; windowInd < SLIDING_WINDOW_SIZE ; windowInd++ ) {
+      TokenImbalanceData memory perBlockData = tokenImbalanceData[token][windowInd];
+      if( perBlockData.lastBlock <= endBlock && perBlockData.lastBlock >= startBlock ) {
+        buyImbalance += int(perBlockData.lastBlockBuyUnitsImbalance);
+      }
+    }
+  }
+
+  function getImbalanceSincePriceUpdate( ERC20 token, uint priceUpdateBlock, uint currentBlock )
+    public view returns(int buyImbalance, int currentBlockImbalance) {
+    buyImbalance = 0;
+    currentBlockImbalance = 0;
+
+    uint64 latestBlock = uint64(0);
+
+    for(uint windowInd = 0 ; windowInd < SLIDING_WINDOW_SIZE ; windowInd++ ) {
+      TokenImbalanceData memory perBlockData = tokenImbalanceData[token][windowInd];
+      if( uint(perBlockData.lastPriceUpdateBlock) != priceUpdateBlock ) continue;
+      if( perBlockData.lastBlock < latestBlock ) continue;
+
+      latestBlock = perBlockData.lastBlock;
+      buyImbalance = perBlockData.totalBuyUnitsImbalance;
+      if( uint(perBlockData.lastBlock) == currentBlock ) {
+        currentBlockImbalance = perBlockData.lastBlockBuyUnitsImbalance;
+      }
+    }
+
+    if( buyImbalance == 0 ) {
+      buyImbalance = getImbalanceInRange( token, priceUpdateBlock, currentBlock );
+    }
+  }
+
+  function getImbalance( ERC20 token, uint priceUpdateBlock, uint currentBlock)
+    public view returns(int totalImbalance, int currentBlockImbalance) {
+
+    int resolution = int(tokenControlInfo[token].minimalRecordResolution);
+
+    (totalImbalance,currentBlockImbalance) = getImbalanceSincePriceUpdate(token,
+                                                                          priceUpdateBlock,
+                                                                          currentBlock );
+    totalImbalance *= resolution;
+    currentBlockImbalance *= resolution;
+  }
+
+
+  function addImbalance( ERC20 token,
+                         int buyAmount,
+                         uint priceUpdateBlock,
+                         uint currentBlock ) internal {
```

Comment:
how is max maxPerBlockImbalance used. and also maxTotalImbalance.
maybe if values were crossed this function should return it as return value.

Reply:
in a different module, we block the tx if it exceeds max. We will verify these are not crossed before calling this function. If they are crossed we return price 0.